### PR TITLE
ARROW-7419: [Python] Support SparseCSCMatrix

### PR DESCRIPTION
**Pull request by rok**
This is to solve [ARROW-7419](https://issues.apache.org/jira/browse/ARROW-7419).
Adding python interface for SparseCSCMatrix.

### CODE REVIEW

**Comment by rok** [comment]
I will look into drying this a bit as it currently just doubles boilerplate.

**Comment by rok** [comment]
This is ready for review.

File: cpp/src/arrow/python/numpy_convert.cc
Comment (pitrou) [inline cpp/src/arrow/python/numpy_convert.cc]: Basically this is the same as `NdarraysToSparseCSRMatrix`, but creating a `SparseCSCIndex` rather than a `SparseCSRIndex`? Perhaps make a common templatized helper?

File: cpp/src/arrow/python/numpy_convert.h
Comment (pitrou) [inline cpp/src/arrow/python/numpy_convert.h]: Is it useful to expose this?

File: python/pyarrow/tensor.pxi
Comment (pitrou) [inline python/pyarrow/tensor.pxi]: "matrix" or "tensor"?

File: python/pyarrow/tensor.pxi
Comment (pitrou) [inline python/pyarrow/tensor.pxi]: Shouldn't you release the GIL here?

File: python/pyarrow/tests/test_serialization.py
Comment (pitrou) [inline python/pyarrow/tests/test_serialization.py]: Why is `data_result[:, 0]` necessary?

File: python/pyarrow/tests/test_serialization.py
Comment (pitrou) [inline python/pyarrow/tests/test_serialization.py]: Don't you check the result type?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this. Here are some comments.

File: cpp/src/arrow/python/numpy_convert.h
Comment (rok) [reply cpp/src/arrow/python/numpy_convert.h]: I'm not sure we currently need it.

File: cpp/src/arrow/python/numpy_convert.cc
Comment (rok) [reply cpp/src/arrow/python/numpy_convert.cc]: Done. Please check.

File: python/pyarrow/tensor.pxi
Comment (rok) [reply python/pyarrow/tensor.pxi]: Matrix indeed.

File: python/pyarrow/tensor.pxi
Comment (rok) [reply python/pyarrow/tensor.pxi]: Done. Please check.

File: python/pyarrow/tests/test_serialization.py
Comment (rok) [reply python/pyarrow/tests/test_serialization.py]: Fixed.

File: python/pyarrow/tests/test_serialization.py
Comment (rok) [reply python/pyarrow/tests/test_serialization.py]: Added.

File: cpp/src/arrow/python/numpy_convert.cc
Comment (pitrou) [inline cpp/src/arrow/python/numpy_convert.cc]: Hmm... I don't think you need to instantiate those explicitly? They're automatically instantiated when compiling the `NdarraysToSparseCSRMatrix` and `NdarraysToSparseCSCMatrix` functions below.

File: python/pyarrow/tests/test_serialization.py
Comment (pitrou) [reply python/pyarrow/tests/test_serialization.py]: Thanks. I also meant the type in the Python sense, e.g. `assert isinstance(pa.SparseCSCMatrix, result)`.
(who knows, perhaps there's a bug and you're getting back a CSR matrix)

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the update. Just two more minor points.

File: cpp/src/arrow/python/numpy_convert.cc
Comment (rok) [reply cpp/src/arrow/python/numpy_convert.cc]: Done.

File: python/pyarrow/tests/test_serialization.py
Comment (rok) [reply python/pyarrow/tests/test_serialization.py]: Added.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you! I'll wait for CI to be green before merging.

**Pull request closed without merging**
